### PR TITLE
start new CtPath Api

### PR DESCRIPTION
**Pull request by petitpre**
Create a new API to select easily some nodes in the Spoon AST.

### CODE REVIEW

File: src/test/java/spoon/test/path/PathTest.java [hunk 1]
```diff
@@ -0,0 +1,110 @@
...
+		Launcher spoon = new Launcher();
+		factory = spoon.createFactory();
+		spoon.createCompiler(
+				factory,
+				SpoonResourceHelper
+						.resources("./src/test/java/spoon/test/path/Foo.java"))
+				.build();
+	}
+	
+	private void equals(CtPath path, CtElement... elements) {
```
Comment (monperrus) [inline src/test/java/spoon/test/path/PathTest.java]: equals(CtPath path, CtElement... elements)
-> 
equals(CtElement root, CtPath path, CtElement... elements)

File: src/main/java/spoon/reflect/path/CtPath.java [hunk 2]
```diff
@@ -0,0 +1,35 @@
...
+	 * @return
+	 */
+	public Collection<CtElement> apply(CtElement... startNode);
+
+	/**
+	 * Search elements that match with this CtPath from the Spoon Factory.
+	 * @param factory
+	 * @return
+	 */
+	public Collection<CtElement> apply(Factory factory);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/CtPath.java]: remove?

File: src/main/java/spoon/reflect/path/CtPathElement.java [hunk 3]
```diff
@@ -0,0 +1,22 @@
...
+
+/**
+ * A single path element from a CtPath.
+ *
+ * @param <P> the type of the queried elements
+ * @param <T> the type of the returned elements
+ */
+public interface CtPathElement<P extends CtElement, T extends CtElement> {
+
+	public Collection<T> getElements(Factory factory);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/CtPathElement.java]: remove?

Comment (monperrus) [inline src/main/java/spoon/reflect/path/CtPathElement.java]: API doc?

File: src/main/java/spoon/reflect/path/CtPathElement.java [hunk 4]
```diff
@@ -0,0 +1,22 @@
+package spoon.reflect.path;
+
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.factory.Factory;
+
+import java.util.Collection;
+
+/**
+ * A single path element from a CtPath.
+ *
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/CtPathElement.java]: in API doc talk about roles

**Comment by monperrus** [comment]
Following our discussion:
- add the API doc to the interfaces?
- remove the *(Factory) methods from the interfaces?
- add tests for path.toString()?

File: src/main/java/spoon/reflect/path/CtPath.java [hunk 5]
```diff
@@ -0,0 +1,23 @@
...
+
+import spoon.reflect.declaration.CtElement;
+
+import java.util.Collection;
+import java.util.List;
+
+/**
+ * A CtPath allow top define the path to a CtElement in the Spoon Model.
+ * <p>
+ * Created by nicolas on 10/06/2015.
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/CtPath.java]: malformed html
remove "created by

File: src/main/java/spoon/reflect/path/CtPath.java [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+ */
+public interface CtPath {
+
+	/**
+	 * Search some element matching this CtPatch from given nodes.
+	 *
+	 * @param startNode
+	 * @return
+	 */
+	public <T extends CtElement> Collection<T> evaluate(Collection<? extends CtElement> startNode);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/CtPath.java]: evaluate on

File: src/main/java/spoon/reflect/path/impl/CtPathElement.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+
+/**
+ * A single path element from a CtPath.
+ *
+ * @param <P> the type of the queried elements
+ * @param <T> the type of the returned elements
+ */
+public interface CtPathElement<P extends CtElement, T extends CtElement> {
+
+	public Collection<T> getElements(Collection<P> roots);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/impl/CtPathElement.java]: api doc?

File: src/main/java/spoon/reflect/path/impl/CtPathElement.java [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+	void addArgument(String key, String value);
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/impl/CtPathElement.java]: api doc with example?

File: src/main/java/spoon/reflect/path/impl/CtPathElement.java [hunk 9]
```diff
@@ -0,0 +1,20 @@
+package spoon.reflect.path.impl;
+
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.factory.Factory;
+
+import java.util.Collection;
+
+/**
+ * A single path element from a CtPath.
+ *
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/impl/CtPathElement.java]: add example?

File: src/main/java/spoon/reflect/path/impl/CtPathImpl.java [hunk 10]
```diff
@@ -0,0 +1,40 @@
...
+import spoon.reflect.declaration.CtElement;
+import spoon.reflect.path.CtPath;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.LinkedList;
+import java.util.List;
+
+/**
+ * Created by nicolas on 12/06/2015.
```
Comment (monperrus) [inline src/main/java/spoon/reflect/path/impl/CtPathImpl.java]: created by

File: src/main/java/spoon/reflect/path/CtPathBuilder.java [hunk 11]
```diff
@@ -0,0 +1,102 @@
...
+import java.util.regex.Pattern;
+
+/**
+ * This builder allow to create some CtPath.
+ * <p>
+ * Some examples:
+ * <p>
+ * To create a CtPath that match with any method in fr.spoon.Launcher:
+ * <p>
+ * new CtPathBuilder().name("fr").name("spoon").name("Launcher").type("method");
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/path/CtPathBuilder.java]:
```
<pre>
{@code
    new CtPathBuilder().name("fr").name("spoon").name("Launcher").type("method");
}
</pre>
```

File: src/main/java/spoon/reflect/path/CtPathRole.java [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+	/**
+	 * Then part of a CtIf 
+	 */
+	THEN("then"),
+	/**
+	 * Else part of a CtIf 
+	 */
+	ELSE("else"),
+	
+	BODY("body");
```
Comment (GerardPaligot) [inline src/main/java/spoon/reflect/path/CtPathRole.java]: Javadoc

**Pull request merged**